### DISCUSSION TITLE
How to group children together in self-relation query

### DISCUSSION THREAD

**Discussion by piotrkulpinski** [prisma/prisma#17242] [Q&A]
Hi,

I'm having a following model in my schema:

```prisma
model Collection {
  id         String       @id @default(cuid())
  title      String
  createdAt  DateTime     @default(now())
  updatedAt  DateTime     @updatedAt
  parent     Collection?  @relation("CollectionChildren", fields: [parentId], references: [id])
  parentId   String?
  children   Collection[] @relation("CollectionChildren")
}
```

What I have is a self-relation of collections. They can have unlimited levels.
Now, I want to query and order and list them so that the final result looks like this:

- Collection 1
- Collection 2
  - Collection 2 Child 1
  - Collection 2 Child 2
  - Collection 2 Child 3
- Collection 3
  - Collection 3 Child 1
    - Collection 3 Child 1 Sub-child 1
  - Collection 3 Child 2
  - Collection 3 Child 3
- Collection 4

The problem is that I have no idea how to query then so that they are ordered by title, while also grouped by parent. 

Is that even possible in Prisma/Postgres?

**Comment by nurul3101** [comment]
Hello @piotrkulpinski 👋 

It would be difficult to get all the children without knowing the depth of the levels of children (as you mentioned that there could be unlimited levels). We have a feature request to support recursive relations here:

- https://github.com/prisma/prisma/issues/3725

If you could leave a 👍 to the feature request, then it would help our product team in prioritising it.

As a workaround, you would need to use [Raw Query](https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access).

The SQL query could look something like this:
```sql
WITH RECURSIVE collection_tree AS (
  SELECT id, title, parentId
  FROM Collection
  WHERE parentId IS NULL
  UNION ALL
  SELECT c.id, c.title, c.parentId
  FROM Collection c
  JOIN collection_tree ct ON ct.id = c.parentId
)
SELECT *
FROM collection_tree
ORDER BY title;
```
This query uses a [recursive CTE](https://www.postgresql.org/docs/current/queries-with.html) to first select the top-level collections (where the parentId is null), and then select all the children of those collections by joining the collection_tree CTE with the Collection table on the parentId. The result is a flattened list of all the collections and their children, ordered by the title field.

Hope this helps!

**Reply by piotrkulpinski** [reply]
Thanks @nurul3101!
I've already upvoted the request and I'm checking the progress on this. 

Unfortunately, the query does not work for me. It returns all collections ordered by title only: 

```
[
  {id: 'clcsxatlf000hst9ikyppemej', title: 'Collection 1', parentId: null}
  {id: 'clcsxayhs000jst9i2o4q0ftv', title: 'Collection 2', parentId: null}
  {id: 'clcsxb2jq000kst9io5dbidu4', title: 'Collection 3', parentId: null}
  {id: 'clcsxb6k5000mst9i1zmu7vtw', title: 'Collection 4', parentId: null}
  {id: 'clcsxbe9y000ost9i2u60h906', title: 'Subcollection 1', parentId: 'clcsxayhs000jst9i2o4q0ftv'}
  {id: 'clcsxblk5000qst9idpxferiw', title: 'Subcollection 2', parentId: 'clcsxayhs000jst9i2o4q0ftv'}
  {id: 'clcsxburg000rst9izb582g6u', title: 'Subcollection 3', parentId: 'clcsxayhs000jst9i2o4q0ftv'}
]
```

Any idea what could be wrong here?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
